### PR TITLE
Fixed let blocks, added percentage block, warn when leaving page

### PR DESCRIPTION
**Pull request by stefan-j**
FunBlocks now warns the user when he tries to leave the page if there are unsaved changes, as per issue #170 

Updated Blockly to handle let variable connecting and disconnecting better. 
The type of all child blocks change when the definition type changes. Child blocks also bump away if there is a type mismatch.
The type changes to polymorphic upon disconnect.
This fixes issue #186. Issue #185 is also fixed, though no additional recursion checking is done.

Added a percentage block. RGBA, Gray and Scaled also now use this percentage block as shadow blocks in the toolbox. This is work for issue #190.
The RGBA block was also fixed.

### CODE REVIEW

File: web/js/blockworld.js [hunk 1]
```diff
@@ -351,8 +356,17 @@ function getWorkspaceXMLText()
 
 function containsUnsavedChanges()
 {
+  var blank = '<xml xmlns="http://www.w3.org/1999/xhtml"></xml>';
+  // Initialize if empty
   if(!lastXML)
-    return false;
+  {
+    lastXML = getWorkspaceXMLText();
```
Comment (cdsmith) [inline web/js/blockworld.js]: This seems wrong.  You are changing lastXML without actually saving anything.  Don't you mean this:

    var blank = '<xml xmlns="http://www.w3.org/1999/xhtml"></xml>';
    return getWorkspaceXMLText() != (lastXML || blank);

Comment (cdsmith) [inline web/js/blockworld.js]: I'm not 100% sure what you're doing here, so I'll merge, and then fix if it doesn't work.

**Comment by stefan-j** [comment]
Thanks ! Your solution is much more elegant. I prefer it.

I can change it, though I have a rather large commit reducing code duplication and I'm unsure how to create a pull request for an entirely separate commit.

**Comment by stefan-j** [comment]
BTW, I tested your version:

```
function containsUnsavedChanges()
{
  var blank = '<xml xmlns="http://www.w3.org/1999/xhtml"></xml>';
  return getWorkspaceXMLText() != (lastXML || blank);
}
```

and it works fine.

**Comment by cdsmith** [comment]
I have made the change locally, and I'll commit it here.

**Pull request merged**